DRIVERS-2884: CSOT avoid connection churn when operations timeout (#1845)
mongodb/specifications: source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.md

Reviewer: Please update the changelog.

Author: Done.

    Pool->>Conn: Draining of pending response from operation (1)
    Conn->>Conn: Update pending read timestamp if bytes read
    alt No error
        Conn->>Conn: Clear pending response state
        Conn->>Pool: Draining of pending response succeeded
        Pool->>Driver: Return connection
    else Error
        alt Non-Timeout Error 
            Conn->>Conn: Close connection
            Conn->>Pool: Draining of pending response failed
            Pool->>Pool: Checkout next connection
        else Timeout Error
            alt Timeout window exceeded
                Conn->>Conn: Close connection
                Conn->>Pool: Draining of pending response failed
                Pool->>Driver: Error
            else Timeout window not exceeded
                Conn->>Pool: Check connection back into pool
                Pool->>Driver: Error
            end
        end
    end
```

##### Pseudocode

This subsection outlines the pseudocode steps for acquiring a connection from the pool.

```text
connection = Null
tConnectionCheckOutStarted = current instant (use a monotonic clock if possible)
emit ConnectionCheckOutStartedEvent and equivalent log message

> checkout_timeout = timeoutMS is set ? remaining computedServerSelectionTimeout : waitQueueTimeoutMS

Reviewer 2: Suggest restructuring to eliminate this variable:

```py
if connection is in "pending response" state:
	if timeoutMS is set:
		draining_timeout = min(remaining computedServerSelectionTiemout, remaining expiration window)
	else: 
		draining_timeout = min(waitQueueTimeoutMS, remaining expiration window)
```

response data is drained and the connection is ready to send the next command.

1. **Persist and update timestamp**: The connection must record the current time immediately after the original timeout.
    This timestamp MUST be updated to the current time whenever any bytes are successfully read, received, or consumed
    while draining of the pending response.
2. **Aliveness check**: If driver uses synchronous read from the socket and the timestamp is older than 3 seconds, the
    driver MUST perform aliveness check either performing a non-blocking read or using the minimal possible timeout to
    check if at least one byte can be read/received. If at least one byte can be read the timestamp must be updated and
    connection SHOULD be drained. If no bytes can be read, the connection MUST be closed.
3. **User-provided timeout**: If a user-provided timeout is specified for the operation, the driver MUST use the minimum
    of (a) the remaining time before the 3 second "pending response" window elapses and (b) the remaining of
    user-provided timeout as the effective timeout for the draining.

> 4. **Default timeout**: If no user-provided timeout is specified, the driver MUST use the minimum of (a) the remaining 3
>     second "pending response" window and (b) the `waitQueueTimeoutMS` (if not supported by the driver remaining of
>     `serverSelectionTimeoutMS` should be used instead) as the effective timeout for the draining.

Reviewer: ```suggestion
4. **Default timeout**: If no user-provided timeout is specified, the driver MUST use the minimum of (a) the remaining 3
    second "pending response" window and (b) remaining timeout for connection checkout.
```

regardless of whether CSOT is enabled or not or waitQueueTimeoutMS is supported, the timeout should always be `min(remaining connection checkout timeout, remaining static timeout)`.  With the existing phrasing, drivers would apply a full `waitQueueTimeoutMS` to every pending read attempt.

        if connection is perished:
          close connection
          connection = Null
        if connection is in "pending response" state:
          tConnectionDrainingStarted = current instant (use a monotonic clock if possible)
          emit PendingResponseStartedEvent and equivalent log message
          try:

>             drain the pending response

Author: @matthewdale Bailey and I agreed to adjust the code very close to what you are suggesting here. With the only difference: `enter WaitQueue` will be under the outer loop. I'll push new version of the pseudocode soon.

>           if error:

Reviewer 3: The mix of exception/error paradigms in this pseudocode block is a bit confusing. We should rewrite this new pseudocode to use try/catch like the existing code.

          except timeout:
            tConnectionDrainingFailed = current instant (use a monotonic clock if possible)

>             emit PendingResponseFailedEvent(reason="timeout", duration = tConnectionDrainingFailed - tConnectionDrainingStarted) and quivalent log message

Reviewer 3: Typo.

```suggestion
            emit PendingResponseFailedEvent(reason="timeout", duration = tConnectionDrainingFailed - tConnectionDrainingStarted) and equivalent log message
```

            if last read timestamp on connection > 3 seconds old
              close connection
            else
              check in the connection

>             connection = Null

Reviewer 3: Thanks for adding the note about the timeout! I checked and confirmed that the Go Driver uses `serverSelectionTimeoutMS` to bound conn checkout currently, but it's easy to miss and more important with this change.

        if connection is perished:
          close connection
          connection = Null

>         if connection is in "pending response" state:

Reviewer: Connection pools are FIFO.  Considering the case where we are successfully reading chunks of data off of the socket but we're timing out before we finish the read (the final else block in your pseudocode), won't checking the connection back into the pool and looping just cause this request to immediately checkout the same connection again (the behavior https://jira.mongodb.org/browse/DRIVERS-3276 is intending to avoid)? 

And regarding https://jira.mongodb.org/browse/DRIVERS-3276: are we convinced that its better to potentially block a number of operations on the slow connection than to block just one?

Author: We can check in the connection back only if the current check out is timed out (either because of CSOT or waitQueueTimeout has elapsed), but "pending response sliding window" has not expired yet. But the "current" operation is timed out anyway, so it should raise an error instead of attempting another connection.

>           drain the pending response

Author: As we have events emission in the current pseudocode, not sure if we need to have pseudocode for "drain the pending response", as it basically "consume the bytes from underling stream/socket and ignore them".

>           drain the pending response

Reviewer: I'd like the pseudocode, it doesn't seem terrible involved to add and it would be nice to codify the calculation of the timeout for the pending read in pseudocode too:

```
read_timeout =  timeoutMS set 
    ? csotMin(timeoutMS, remaining static timeout) 
    : waitQueueTimeoutMS set 
        ? csotMin(waitQueueTimeoutMS, remaining static timeout) 
        : remaining static timeout
```

          if error:

>             if non-timeout error:

Reviewer: Pseudocode needs to include new event and logging emissions.

   */
  connectionId: int64;
}

/**
 *  Emitted when the connection is attempting to drain of the pending response.
 */
interface PendingResponseStarted {
  /**
   *  The ServerAddress of the Endpoint the pool is attempting to connect to.
   */
  address: string;

  /**
   *  The ID of the Connection.
   */
  connectionId: int64;

  /**
   *  The driver-generated request ID of the operation that caused the pending response state.
   */

>   requestId: int64;

Reviewer: Seems odd that the events emitted for an operation include a requestId that corresponds to the request that made the connection `pending response`.  What is the value in this datapoint?

If we decide to keep it - do you think a more precise name would be beneficial?  `requestId` is ambiguous - users could easily assume it refers to the request that is reading the pending response, not the operation that made the connection pending.

Author: The idea of reporting the original requestId is to keep tracking how long/how many draining attempt were made before success or failure. Honestly I would prefer to have BOTH, current request Id and "original timed out requestId", but it could make sense only if other check out event had "current requestId" field, which is not there =(

Reviewer: Yeah, fine to keep it - thoughts on workshopping the name to make it clearer?

   */
  connectionId: int64;
}

/**
 *  Emitted when the connection being checked out is attempting to read and 
 *  discard a pending server response.
 */
interface PendingResponseStarted {
  /**
   *  The ServerAddress of the Endpoint the pool is attempting to connect to.
   */
  address: string;

  /**
   *  The ID of the Connection.
   */
  connectionId: int64;

  /**
   *  The driver-generated request ID associated with the network timeout.
   */

>   requestId: int64;

Author: Current approach from the log reader prospective will looks like: there are many operations, there are many checkout attempts, some of them faster, some slower, and also there are draining of connections happening. No way to build a relation between connection checkout duration and connections draining. Slower checkouts might be caused by the draining or simply waiting. If we have `requestId` and `originalRequestId` (need better name sure) - then reader can see, we have a longer checkout for THIS `requestId`, because there were some connection draining for that `originalRequestId`.

5.0. The advantage of using Background Thread to manage perished connections is that it will work regardless of
environment setup.

> 

Reviewer 4: Could we add a rationale question to cover the motivation for this feature? Like "Why introduce the draining pending responses?" where the answer is to reduce connection churn in cases when the configured maxTimeMS does not allow enough time for the driver to read the MaxTimeMSExpired error and cases where the server or network delays the response.

Author: Done.
